pipelines.md

```diff
@@ -0,0 +1,358 @@
+    |> line([1, 0])
+    ...
+    |> close()
+  }
+}
+```
+
+When used in a pipeline, the input to `for` is passed into `expr_2` as the receiver for the first iteration. The result of that execution is passed into the next iteration and so forth until the final result is returned as the value of the for expression. I.e., it is a kind of reduce or fold operation. E.g.,
+
+```
+sketch
+  |> for [0..n] as i
+    |> drawOneSectorOfGear(i)
+  |> extrude(...)
+```
+
+Here, `sketch` is the receiver of `drawOneSectorOfGear` on the first iteration, and the result of that is the receiver on the next. The final result is the extrusion of the inital sketch with `n` sectors.
+
+Example with block:
+
+```
+sketch
+  |> for [0..n] as i |> {
+    x = sin(i)
+    y = cos(i)
+    drawOneSectorOfGear(x, y)
+  }
+  |> extrude(...)
+```
+
+Note that only a single pipeline stage is repeated (thus the indentation in the first example). To apply longer pipelines, use a block, e.g.,:
+
+```
+sketch
+  |> for [0..n] as i |> {
+    drawOneSectorOfGear(i)
+      |> foo()
+  }
+  |> extrude(...)
+```
+
+Tagging using `as` works as normal within a block, tagging the whole block (or the whole single expression, if there is no block), produces an array of tags. E.g.,
+
+```
+sketch
+  |> for [0..n] as i
+    |> drawOneSectorOfGear(i) as x
+  |> extrude(...)
+
+// or
+
+sketch
+  |> for [0..n] as i |> {
+    x = sin(i)
+    y = cos(i)
+    drawOneSectorOfGear(x, y)
+  } as x
+  |> extrude(...)
+```
+
+In both cases, `x` has type [T] if `drawOneSectorOfGear` returns a `T`.
+
+Extension: Assumes we can use `export as` to export a variable from a block. Tagging within a for loop produces a somewhat special variable: it refers to the current iteration's tag within the loop and an array of tags outside the loop. E.g.,
+
+```
+sketch
+  |> for [0..n] as i |> {
+    drawOneSectorOfGear(i) export as x
+      |> foo(x) // x has type T
+  }
+  |> bar(x)     // x has type [T]
+  |> extrude(...)
+```
+
+### Alternative syntax
+
+Use `in` instead of `|>` in the for loop syntax. Examples:
+
+```
+x = [0..5] as i in i * 2
+
+sketch
+  |> for [0..n] as i in
+    drawOneSectorOfGear(i)
+  |> extrude(...)
+
+sketch
+  |> for [0..n] as i in {
+    drawOneSectorOfGear(i)
+      |> foo()
+  }
+  |> extrude(...)
+```
+
+I think this works better in the block case, but not so well in the case with a single expression.
+
+Or we could make the braced block mandatory rather than optional with a single expression:
+
+```
+x = [0..5] as i { i * 2 }
+
+sketch
+  |> for [0..n] as i {
+    drawOneSectorOfGear(i)
+  }
+  |> extrude(...)
+
+sketch
+  |> for [0..n] as i {
+    drawOneSectorOfGear(i)
+      |> foo()
+  }
+  |> extrude(...)
+```
+
+Or we could have a mandatory braced block *and* either `|>` or `in`.
+
+And/or, we could use `for i in [...]` rather than `for [...] as i`. This is closer to other PLs and reads better, however, it does not reuse the `as` keyword and so requires a new keyword.
```

Comment:
Using a new keyword doesn't bother me. I guess it can make the parser implementation more complicated because it might mean that we need to do more contextual parsing. (For `import`, we ended up checking the following character in the tokenizer -- `import ` vs. `import(` -- and that felt dirty. Would have been cleaner, IMO, to check the following _token_ in the parser.) But I think it's worth it for simple things like this.

I think the reason to go with `as i` is so that users have one fewer thing to learn. When they learn it in `for` loops, it's training them to use it elsewhere. Then again, anyone who's programmed before gets `for i in`, so that could probably work too. Personally, I find `as i` more elegant and consistent as a design.